lib/Dialect/LinalgExt/Transforms/Fusion.cpp

```diff
@@ -0,0 +1,121 @@
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/OperationSupport.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/TilingInterface.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::iree_compiler::IREE::LinalgExt;
+
+/// Returns the `producerOp` loop dimensions that map to the `opResult` shape
+/// dimensions.
+static FailureOr<SmallVector<int64_t>>
+getIteratorIndicesTiedToOpResult(linalg::LinalgOp producerOp,
+                                 OpResult opResult) {
+  assert(opResult.getOwner() == producerOp && "expected producer result");
+  AffineMap indexingMap = producerOp.getTiedIndexingMapForResult(opResult);
+  if (!indexingMap.isProjectedPermutation())
+    return failure();
+  SmallVector<int64_t> tiedIndices;
+  for (AffineExpr expr : indexingMap.getResults())
+    tiedIndices.push_back(expr.cast<AffineDimExpr>().getPosition());
+  return tiedIndices;
+}
+
+/// Returns the `producerOp` loop ranges tiled to the `sliceOp` shape or failure
+/// if the computation fails.
+static FailureOr<SmallVector<Range>>
+getTiledIterationDomain(OpBuilder &b, linalg::LinalgOp producerOp,
+                        tensor::ExtractSliceOp sliceOp) {
+  // Set the iteration domain to the ranges before tiling.
+  SmallVector<Range> tiledIterationDomain =
+      cast<TilingInterface>(producerOp.getOperation()).getIterationDomain(b);
+
+  // Update the tiled dimensions given the `sliceOp` offset and sizes.
+  FailureOr<SmallVector<int64_t>> tiedIndices =
+      getIteratorIndicesTiedToOpResult(producerOp,
+                                       sliceOp.source().cast<OpResult>());
+  if (failed(tiedIndices))
+    return failure();
+  for (auto it : llvm::zip(tiedIndices.getValue(), sliceOp.getMixedOffsets()))
+    tiledIterationDomain[std::get<0>(it)].offset =
+        getValueOrCreateConstantIndexOp(b, sliceOp->getLoc(), std::get<1>(it));
+  for (auto it : llvm::zip(tiedIndices.getValue(), sliceOp.getMixedSizes()))
+    tiledIterationDomain[std::get<0>(it)].size =
+        getValueOrCreateConstantIndexOp(b, sliceOp->getLoc(), std::get<1>(it));
+
+  return tiledIterationDomain;
+}
+
+FailureOr<SmallVector<linalg::LinalgOp>>
+mlir::iree_compiler::IREE::LinalgExt::LinalgExtFusionPattern::
+    returningMatchAndRewrite(linalg::LinalgOp producerOp,
+                             PatternRewriter &rewriter) const {
+  if (!producerOp.hasTensorSemantics() && producerOp->getNumResults() != 1)
+    return failure();
+
+  // Search the producer slices accessed within the containing operation.
+  SmallVector<tensor::ExtractSliceOp> sliceOps;
+  for (Operation *user : producerOp->getUsers()) {
+    auto sliceOp = dyn_cast<tensor::ExtractSliceOp>(user);
+    if (!sliceOp)
+      continue;
+    if (sliceOp->getParentOp() != containingOp)
+      continue;
+    sliceOps.push_back(sliceOp);
+  }
+  // Check for a non-empty list of fusion opportunities.
+  if (sliceOps.empty())
```

Comment:
Fun extension for the future that I realized when writing a design doc, we can relax this and get other forms of fusion than just producer-consumer by simply adding the result slices to the terminator :)

Reply:
right the dummy op is not needed. Although tiling a dummy op may still be the most convenient way to get the slices.